### ISSUE TITLE
Limit websocket connection to just channels that are active

### ISSUE THREAD

**Issue by talk2MeGooseman** [talk2MeGooseman/stream_closed_captioner_phoenix#76]

**Comment by talk2MeGooseman** [comment]
## Implementation Plan

> _Generated from issue analysis — comment with feedback to revise_

### Understanding

The issue asks that WebSocket connections be limited to channels where the streamer is actively captioning. The codebase already has `UserTracker.channel_active?/1` (returns `true` if `last_publish` is within 300 seconds) — but this check is never consulted at connection time. There are two connection paths that need gating:

1. **Absinthe GraphQL subscription `new_twitch_caption`** — this is what the Twitch extension uses to receive captions and currently has no activity gate. This is the primary target.
2. **`CaptionsChannel` track-on-join** — `track_user/1` currently registers the streamer in `UserTracker` with empty metadata (`%{}`), meaning `channel_active?` returns `false` until the first `"publishFinal"` or `"active"` push. This creates a race window that must be closed first so the subscription gate works correctly from the moment a streamer connects their dashboard.

The Phoenix Channel join itself (`join/3`) already uses `authorized?/2` to limit connections to the channel owner, so the streamer's own connection doesn't need a new activity gate.

---

### Tests to write first

#### Unit tests — `test/stream_closed_captioner_phoenix_web/channels/captions_channel_test.exs`

1. **"track_user sets last_publish on join"**
   - Setup: `subscribe_and_join/3` for a user's own channel
   - Assert: `UserTracker.get_by_key("active_channels", uid)` returns metadata with `last_publish` set close to `System.system_time(:second)`
   - Why: Currently `track_user/1` passes `%{}`, so `last_publish` is absent

2. **"channel_active? returns true immediately after streamer joins their channel"**
   - Setup: `subscribe_and_join/3` succeeds
   - Assert: `UserTracker.channel_active?(uid) == true`
   - Why: Integration proof that the race condition (join → activity) is closed

#### Unit tests — Absinthe subscription (find or create under `test/stream_closed_captioner_phoenix_web/`)

3. **"new_twitch_caption subscription succeeds when channel is active"**
   - Setup: `UserTracker.track(self(), "active_channels", channel_id, %{last_publish: System.system_time(:second)})` in the test setup
   - Action: push GraphQL subscription request for `newTwitchCaption(channelId: channel_id)`
   - Assert: subscription config returns `{:ok, topic: channel_id}` (no error)
   - Add `on_exit` cleanup to `UserTracker.untrack`

4. **"new_twitch_caption subscription is rejected when channel is inactive (stale timestamp)"**
   - Setup: `UserTracker.track` with `last_publish` set to `System.system_time(:second) - 400` (beyond the 300s timeout)
   - Action: push subscription request for that channel_id
   - Assert: config returns `{:error, _}` and subscription is rejected

5. **"new_twitch_caption subscription is rejected when channel has never been active"**
   - Setup: no `UserTracker` entry for the channel_id at all
   - Action: push subscription request
   - Assert: config returns `{:error, _}`
   - Why: ensures the `nil`/missing case is handled gracefully, not just the stale case

#### Integration test — `test/stream_closed_captioner_phoenix_web/channels/captions_channel_test.exs` or a subscription test

6. **"caption only reaches subscriber after streamer connects"**
   - Step A: attempt `newTwitchCaption` subscription before streamer joins → assert error/rejection
   - Step B: streamer `subscribe_and_join/3` for their channel
   - Step C: attempt `newTwitchCaption` subscription again → assert success
   - Step D: streamer pushes `"publishFinal"` with a caption
   - Step E: assert the subscription receives the `newTwitchCaption` event

---

### Implementation order

**Cycle 1 — Close the race condition (drive `track_user` to set `last_publish` on join)**
1. Write tests #1 and #2 → **red** (metadata is `%{}`, `channel_active?` returns `false`)
2. In `captions_channel.ex`, change `track_user/1`:
   ```elixir
   # Before
   UserTracker.track(self(), "active_channels", uid, %{})

   # After
   UserTracker.track(self(), "active_channels", uid, %{last_publish: System.system_time(:second)})
   ```
3. Tests #1 and #2 → **green**

**Cycle 2 — Gate the Absinthe subscription**
1. Write tests #3, #4, and #5 → **red** (no check in subscription config)
2. Locate the `new_twitch_caption` subscription definition in `schema.ex` (or its import). In the `config` function:
   ```elixir
   config fn args, _info ->
     if StreamClosedCaptionerPhoenixWeb.UserTracker.channel_active?(args.channel_id) do
       {:ok, topic: args.channel_id}
     else
       {:error, "channel is not active"}
     end
   end
   ```
3. Tests #3–5 → **green**

**Cycle 3 — Integration**
1. Write test #6 → expect red, fix any integration issues found
2. Run `mix test` for the full suite; pay special attention to:
   - `test/stream_closed_captioner_phoenix/jobs/send_chat_reminder_test.exs` — also calls `UserTracker.channel_active?`
   - Any showcase-related tests that call `recently_active_channels/0` — fixing `track_user/1` means channels now appear active as soon as the streamer connects (before first caption), which may affect the showcase test expectations

---

### Files to change

| File | Change |
|------|--------|
| `lib/stream_closed_captioner_phoenix_web/channels/captions_channel.ex` | `track_user/1`: pass `%{last_publish: System.system_time(:second)}` instead of `%{}` |
| `lib/stream_closed_captioner_phoenix_web/graphql/schema.ex` (or where `new_twitch_caption` is defined) | Add `UserTracker.channel_active?` check inside the subscription's `config` function |
| `test/stream_closed_captioner_phoenix_web/channels/captions_channel_test.exs` | Add tests #1, #2, and #6 |
| `test/stream_closed_captioner_phoenix_web/graphql/` (find existing or create new subscription test file) | Add tests #3, #4, #5 |

---

### Edge cases and considerations

1. **Chicken-and-egg (join → activity gap):** Covered by Cycle 1. Without fixing `track_user/1` first, the subscription gate would reject the very moment viewers try to connect right after a streamer joins. Tests #1 and #2 prove this is closed.

2. **Existing subscriptions when a channel goes inactive:** Once subscribed, the Absinthe subscription stays open — the gate only rejects new subscription requests. Viewers already watching will naturally stop receiving events when the streamer stops publishing. This is acceptable behavior that matches current UX. Document this in the PR.

3. **`channel_id` ↔ `user.uid` mapping:** `UserTracker` keys channels by `user.uid` (Twitch broadcaster string ID). The Twitch extension JWT contains `channel_id` (also the Twitch broadcaster ID). These should be the same value — verify by checking the subscription's `config` resolver context and confirming `args.channel_id` is the Twitch user ID string. **This is the highest-risk unknown; confirm before implementing Cycle 2.**

4. **Side effects on `recently_active_channels/0` callers:** The showcase page and `SendChatReminder` job use `UserTracker.recently_active_channels/0`, which filters by `last_publish`. Fixing `track_user/1` to set `last_publish` on join means channels appear "active" from the moment the streamer opens their dashboard, even before speaking. The showcase will show more channels. Evaluate whether this is desired or if a separate "is captioning" signal is needed.

5. **Silent frontend failures:** `captions.js` (dashboard) only logs `console.debug` on join error; the extension's subscription observer has no error callback. If we reject subscriptions, viewers get no feedback. Consider adding visible error handling in a follow-up — out of scope for this issue but worth noting in the PR.

6. **Server restart resets all activity:** `UserTracker` is in-memory (Phoenix.Tracker). After a restart, all channels are inactive until streamers reconnect. This is correct behavior (no stale "active" state), but could surprise integrators. No code change needed; document in the PR.

---

### Open questions

1. **Does `args.channel_id` in the subscription config equal `user.uid` in `UserTracker`?** Confirm by tracing how the Twitch extension JWT's `channel_id` is extracted and passed as the subscription arg, and compare to what `user.uid` contains in the User schema.

2. **Should the Phoenix Channel join (`join/3`) also be gated?** Currently, `authorized?/2` already limits channel joins to the owner (the streamer), so non-owners cannot join. A separate `channel_active?` check at the join level would only affect the streamer reconnecting after a 5+ minute idle — which is probably undesirable. Confirm with the issue author.

3. **Should existing subscriptions be forcibly terminated when a channel goes inactive?** The proposed change only rejects new subscriptions. Proactive disconnection of current viewers when a streamer goes offline is a harder problem (requires broadcasting a termination event through Absinthe). If desired, it's a follow-up issue.

4. **Is 300 seconds the right active timeout for the subscription gate?** The `@active_time_out 300` constant in `UserTracker` also drives the showcase and reminder logic. If different timeouts are needed for different use cases, consider making it a parameter rather than changing the constant.


---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by talk2MeGooseman** [comment]
## Implementation Plan _(revised)_

> _Revised to incorporate additional risk analysis findings_

### Revision notes

Three things changed from the initial plan:

1. **Test setup constraint added**: `UserTracker` is a supervised `GenServer`, not an Elixir behaviour, so it **cannot be mocked with Mox**. All channel tests that involve `UserTracker` state must use the real tracker and run with `async: false` (matching the existing `UserTrackerTest` pattern). The test descriptions now reflect this explicitly.

2. **`ActivePresence` disambiguation**: The codebase has two parallel activity-tracking systems — `UserTracker` (300s timeout) and `ActivePresence` (700s timeout). The implementation must use `UserTracker` exclusively, since it is the one actively used in production code (`SendChatReminder`, `ShowcaseController`). `ActivePresence` is largely unused and should not be consulted for the gate.

3. **New edge case test added (#7)**: The join-time check does not protect against a connection that was accepted at join and then goes idle for 300+ seconds. Test #7 documents the expected behavior for this scenario (idle connection can still send captions — this is acceptable and consistent with current behavior) so the decision is explicit rather than implicit.

---

### Understanding

The issue asks that WebSocket connections be limited to channels where the streamer is actively captioning. The codebase already has `UserTracker.channel_active?/1` (returns `true` if `last_publish` is within 300 seconds) — but this check is never consulted at connection time. There are two gaps to close:

1. **Absinthe GraphQL subscription `new_twitch_caption`** — this is what the Twitch extension uses to receive captions and currently has no activity gate. This is the primary target.
2. **`CaptionsChannel.track_user/1` sets empty metadata on join** — `track_user/1` currently calls `UserTracker.track` with `%{}`, so `channel_active?` returns `false` until the first `"publishFinal"` or `"active"` push. This creates a race window that must be closed first so the subscription gate works correctly from the moment a streamer connects.

The Phoenix Channel join itself (`join/3`) is already limited to the channel owner by `authorized?/2` and does not need a new activity gate — only the streamer can join their own channel, and they need to join in order to _become_ active.

---

### Tests to write first

> All tests touching `UserTracker` must run with `async: false` — `UserTracker` is a supervised GenServer (not a behaviour), cannot be mocked with Mox, and its global state is not sandbox-isolated. Follow the pattern in the existing `UserTrackerTest` and `CaptionsChannelTest`, which already enforce this.

#### Unit tests — `test/stream_closed_captioner_phoenix_web/channels/captions_channel_test.exs`

1. **"track_user sets last_publish on join"**
   - Setup: `subscribe_and_join/3` for a user's own channel; add `on_exit` to `UserTracker.untrack`
   - Assert: `UserTracker.get_by_key("active_channels", uid)` returns metadata with `last_publish` near `System.system_time(:second)`
   - Currently failing because `track_user/1` passes `%{}`

2. **"channel_active? returns true immediately after streamer joins"**
   - Setup: `subscribe_and_join/3` succeeds
   - Assert: `UserTracker.channel_active?(uid) == true`
   - Integration proof that the join-to-activity race is closed

3. **"long-idle connection can still send captions after 300s timeout"** _(documents accepted behavior)_
   - Setup: `subscribe_and_join/3`; manually update `UserTracker` metadata with a stale `last_publish` (e.g., `System.system_time(:second) - 400`)
   - Action: push `"publishFinal"` on the existing socket
   - Assert: caption is processed normally and `UserTracker` `last_publish` is refreshed
   - Why: the join-time gate does not disconnect existing connections. This test documents that post-join inactivity is _not_ enforced by this issue and that a future issue would be needed for that.

#### Unit tests — Absinthe subscription (find or create under `test/stream_closed_captioner_phoenix_web/`)

4. **"new_twitch_caption subscription succeeds when channel is active"**
   - Setup: `UserTracker.track(self(), "active_channels", channel_id, %{last_publish: System.system_time(:second)})` in test; add `on_exit` to untrack
   - Action: push GraphQL subscription for `newTwitchCaption(channelId: channel_id)`
   - Assert: config returns `{:ok, topic: channel_id}` (no error)

5. **"new_twitch_caption subscription is rejected when channel is inactive (stale timestamp)"**
   - Setup: `UserTracker.track` with `last_publish` set to `System.system_time(:second) - 400`
   - Action: push subscription request
   - Assert: config returns `{:error, _}`

6. **"new_twitch_caption subscription is rejected when channel has no UserTracker entry"**
   - Setup: ensure no `UserTracker` entry exists for the channel_id
   - Action: push subscription request
   - Assert: config returns `{:error, _}`
   - Why: tests the `nil`/missing case separately from the stale case to ensure the `nil` guard doesn't crash

#### Integration test

7. **"caption only reaches subscriber after streamer connects"**
   - Step A: attempt `newTwitchCaption` subscription before streamer joins → assert error
   - Step B: streamer `subscribe_and_join/3` for their channel
   - Step C: attempt subscription again → assert success
   - Step D: streamer pushes `"publishFinal"` with a caption
   - Step E: assert subscription receives the `newTwitchCaption` event

---

### Implementation order

**Cycle 1 — Close the join-to-activity race (drive `track_user` to set `last_publish` on join)**
1. Write tests #1 and #2 → **red**
2. In `captions_channel.ex`, change `track_user/1`:
   ```elixir
   # Before
   UserTracker.track(self(), "active_channels", uid, %{})

   # After
   UserTracker.track(self(), "active_channels", uid, %{last_publish: System.system_time(:second)})
   ```
3. Tests #1 and #2 → **green**
4. Write test #3 → should be **green immediately** (documents existing behavior; if it fails, the `handle_in` pipeline has a new regression)

**Cycle 2 — Gate the Absinthe subscription**
1. Write tests #4, #5, and #6 → **red**
2. In the `new_twitch_caption` subscription `config` function:
   ```elixir
   config fn args, _info ->
     if StreamClosedCaptionerPhoenixWeb.UserTracker.channel_active?(args.channel_id) do
       {:ok, topic: args.channel_id}
     else
       {:error, "channel is not active"}
     end
   end
   ```
3. Tests #4–6 → **green**

**Cycle 3 — Integration and regression check**
1. Write test #7 → expect green if Cycles 1–2 are correct; investigate any failures
2. Run `mix test` for the full suite; watch especially:
   - `send_chat_reminder_test.exs` — also calls `UserTracker.channel_active?`
   - Any showcase tests that call `recently_active_channels/0` — Cycle 1 means channels now appear "active" as soon as the streamer opens their dashboard (before first caption), which may change showcase test expectations
3. Run `mix lint` and `mix security` before marking done

---

### Files to change

| File | Change |
|------|--------|
| `lib/stream_closed_captioner_phoenix_web/channels/captions_channel.ex` | `track_user/1`: pass `%{last_publish: System.system_time(:second)}` instead of `%{}` |
| `lib/stream_closed_captioner_phoenix_web/graphql/schema.ex` (or wherever `new_twitch_caption` subscription is defined) | Add `UserTracker.channel_active?` check in the subscription's `config` fn |
| `test/stream_closed_captioner_phoenix_web/channels/captions_channel_test.exs` | Add tests #1, #2, #3, and #7; ensure `async: false` is set at the module level |
| `test/stream_closed_captioner_phoenix_web/graphql/` (find existing or create subscription test file) | Add tests #4, #5, #6; `async: false` required |

---

### Edge cases and considerations

1. **`UserTracker` vs. `ActivePresence`**: The codebase has two parallel tracking systems. Use `UserTracker` only — it has the 300s timeout and is the one already used in `SendChatReminder` and `ShowcaseController`. `ActivePresence` (700s timeout) is largely unused; do not introduce it here.

2. **`UserTracker` cannot be mocked**: It is a supervised GenServer started in `Application`, not a behaviour. Tests must call the real `UserTracker` and run with `async: false`. Add `on_exit` hooks to `untrack` any test-seeded entries so state doesn't leak between tests (see existing `UserTrackerTest` pattern at line 8).

3. **Existing connections past the 300s timeout**: A connection accepted at join can still send captions after going idle for 300+ seconds. The join-time gate does not close this window. Test #3 documents this as accepted behavior. If post-join enforcement is ever needed, it requires a separate per-message check or a proactive disconnect on `stream.offline` — that is a follow-up issue.

4. **`channel_id` ↔ `user.uid` mapping**: The Twitch extension JWT `channel_id` field and `user.uid` in the Phoenix app should both be the Twitch broadcaster string ID. Verify by tracing how the subscription `config` receives `args.channel_id` and comparing to what `UserTracker.update` uses (`user.uid` from `socket.assigns.current_user.uid`). If they differ, a DB lookup via `Accounts` will be needed before calling `channel_active?`. **Highest-risk unknown — verify before implementing Cycle 2.**

5. **Showcase side effect**: Fixing `track_user/1` to set `last_publish` on join means a channel registers as "active" as soon as the streamer opens their dashboard, before they speak. This may surface more channels in the showcase display. Evaluate whether the showcase should use "has published" rather than "has joined" as its definition of active.

6. **Silent frontend failures**: `captions.js` only logs `console.debug` on join error; the extension's GraphQL subscription observer has no error callback. If subscription rejection should be surfaced to the viewer, frontend error handling is a follow-up — out of scope here, but note it in the PR.

7. **Server-restart resets activity**: `UserTracker` is in-memory. After a restart, all channels are inactive until streamers reconnect. This is correct behavior (no stale "active" state persists) but worth documenting in the PR for integrators.

---

### Open questions

1. **Does `args.channel_id` equal `user.uid`?** Confirm before implementing the subscription gate. If the Twitch JWT's `channel_id` is a different format or field than `user.uid`, the gate will silently pass or block everything.

2. **Should the Phoenix Channel `join/3` also gate on activity?** Currently `authorized?/2` already limits joins to the channel owner. An additional `channel_active?` check at join would block the streamer from reconnecting after a 5-minute idle — which is almost certainly wrong. Confirm the intent is to gate *consumer subscriptions only*, not the streamer's own connection.

3. **Should the gate use `stream.offline` EventSub instead of (or in addition to) `UserTracker`?** The codebase already handles `stream.offline` webhooks in `webhooks_controller.ex`. If the intent is "only active when the Twitch stream itself is live," the `UserTracker` heuristic (300s since last caption) may not match that. Clarify the definition of "active" with the issue author.

4. **Should existing subscriptions be terminated when a channel goes inactive?** The proposed change only rejects new subscriptions. Proactively disconnecting current viewers when a streamer's channel goes inactive is a harder problem (requires broadcasting through Absinthe) and is a follow-up if desired.


---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by talk2MeGooseman** [comment]
Does args.channel_id equal user.uid? Confirm before implementing the subscription gate. If the Twitch JWT's channel_id is a different format or field than user.uid, the gate will silently pass or block everything.

Yes

Should the Phoenix Channel join/3 also gate on activity? Currently authorized?/2 already limits joins to the channel owner. An additional channel_active? check at join would block the streamer from reconnecting after a 5-minute idle — which is almost certainly wrong. Confirm the intent is to gate consumer subscriptions only, not the streamer's own connection.

No

Should the gate use stream.offline EventSub instead of (or in addition to) UserTracker? The codebase already handles stream.offline webhooks in webhooks_controller.ex. If the intent is "only active when the Twitch stream itself is live," the UserTracker heuristic (300s since last caption) may not match that. Clarify the definition of "active" with the issue author.

Active means they have turn on captions on the dashboard, but it should be more forgiving than the current heuristic, so long as the have turn on captions on the dashboard that will be considered "active"

Should existing subscriptions be terminated when a channel goes inactive? The proposed change only rejects new subscriptions. Proactively disconnecting current viewers when a streamer's channel goes inactive is a harder problem (requires broadcasting through Absinthe) and is a follow-up if desired.

Yes, so long as it doesn't add to the complexity

**Comment by talk2MeGooseman** [comment]
## Implementation Plan _(revised)_

> _Revised in response to: "Does args.channel_id equal user.uid? ... Yes ... Active means they have turn on captions on the dashboard..."_

### Revision notes

Four things changed from the previous plan:

1. **Open question #1 resolved (channel_id == uid confirmed):** `args.channel_id` in the subscription config equals `user.uid` in `UserTracker`. The subscription gate can call `UserTracker.channel_connected?(args.channel_id)` directly with no DB lookup.

2. **Open question #2 resolved (join/3 gate confirmed No):** Only the Absinthe consumer subscription is gated — not the streamer's own `CaptionsChannel` join. Removed from open questions.

3. **"Active" definition changed — subscription gate predicate updated:** The issue owner clarified: "active means they have turned on captions on the dashboard." This means the gate should check **whether the streamer is currently connected to their `CaptionsChannel`** (has any `UserTracker` entry), NOT whether they've published within the last 300 seconds (`channel_active?`). A streamer with their dashboard open but silent for 5 minutes should still allow viewers to subscribe. This requires:
   - Adding `UserTracker.channel_connected?/1` (checks `get_by_key/2` returns a non-empty list)
   - Replacing `channel_active?` with `channel_connected?` in the subscription `config` fn
   - Old test "subscription rejected when stale timestamp" is **removed** and replaced by "subscription **succeeds** when streamer is connected but hasn't published in 300s" — explicitly documenting the relaxed gate semantics

4. **Subscription termination on disconnect deferred:** The issue owner asked for existing subscriptions to be terminated when the streamer disconnects. Absinthe's public API (`Absinthe.Subscription.unsubscribe/2`) operates per subscription-ID, not per topic — there is no public API to close all subscribers for a given topic simultaneously. Doing this would require using Absinthe's private internal ETS tables or maintaining a separate registry mapping each channel to its active subscription IDs. Both add significant complexity. Per the author's qualifier ("so long as it doesn't add to the complexity"), proactive subscriber termination is deferred to a follow-up issue. The practical impact is low: once the streamer disconnects, `CaptionsChannel` stops publishing, so existing viewer connections go idle with no data incoming.

---

### Understanding

The issue asks that WebSocket connections be limited to channels where the streamer is actively captioning. The codebase already has `UserTracker` (a `Phoenix.Tracker`-based GenServer) that tracks which channels are open — but this is never consulted when a viewer establishes an Absinthe GraphQL subscription.

The issue owner has clarified "active" means the streamer has their captions dashboard open (i.e., is currently connected to `CaptionsChannel`), not that they've published recently. Two gaps to close:

1. **`UserTracker` needs a `channel_connected?/1` function** — the existing `channel_active?/1` checks a 300s recency window (used by `SendChatReminder` and the showcase); the subscription gate needs a simpler "is there any entry at all" predicate.
2. **The `new_twitch_caption` subscription `config` fn needs an activity check** — it currently accepts any `channel_id` unconditionally (`{:ok, topic: args.channel_id}`).

A secondary fix: **`CaptionsChannel.track_user/1` currently registers with empty metadata (`%{}`)**, so `channel_active?` returns `false` until the first caption is published. Setting `last_publish` on join fixes this for `SendChatReminder` and the showcase.

---

### Tests to write first

> All tests touching `UserTracker` must run without `async: true` — `UserTracker` is a supervised GenServer, not a behaviour, and cannot be mocked with Mox. Its global state is not sandbox-isolated. Follow the existing `UserTrackerTest` pattern: add `on_exit(fn -> UserTracker.untrack(self(), "active_channels", uid) end)` for every test-seeded entry.

#### Unit tests — `test/stream_closed_captioner_phoenix_web/channels/user_tracker_test.exs`

**Test 1: "channel_connected? returns true when a UserTracker entry exists"**
- Setup: `UserTracker.track(self(), "active_channels", "uid123", %{last_publish: 0})`; `on_exit` to untrack
- Assert: `UserTracker.channel_connected?("uid123") == true`

**Test 2: "channel_connected? returns false when no UserTracker entry exists"**
- Setup: call `UserTracker.untrack(self(), "active_channels", "nonexistent")` first to ensure no entry
- Assert: `UserTracker.channel_connected?("nonexistent") == false`

**Test 3: "channel_connected? returns true even when last_publish is stale (>300s ago)"**
- Setup: `UserTracker.track(self(), "active_channels", "uid456", %{last_publish: System.system_time(:second) - 400})`; `on_exit` to untrack
- Assert: `UserTracker.channel_connected?("uid456") == true` **AND** `UserTracker.channel_active?("uid456") == false`
- Why: explicitly documents the semantic split between `channel_connected?` (gate predicate) and `channel_active?` (SendChatReminder / showcase predicate); prevents future developers from accidentally swapping them

#### Unit tests — `test/stream_closed_captioner_phoenix_web/channels/captions_channel_test.exs`

**Test 4: "track_user sets last_publish on join"**
- Setup: `subscribe_and_join/3` for a user's own channel; `on_exit` to `UserTracker.untrack`
- Assert: `UserTracker.get_by_key("active_channels", uid)` returns a list whose first element's metadata contains `last_publish` within 2 seconds of `System.system_time(:second)`
- Currently failing because `track_user/1` passes `%{}`

**Test 5: "channel_active? returns true immediately after streamer joins"**
- Setup: `subscribe_and_join/3` succeeds (requires Test 4 to be passing first)
- Assert: `UserTracker.channel_active?(uid) == true`
- Integration proof that the join-to-activity race is closed for `SendChatReminder`

**Test 6: "long-idle connection can still send captions after 300s" (documents accepted behavior)**
- Setup: `subscribe_and_join/3`; manually call `UserTracker.update` to set a stale `last_publish` (e.g., `System.system_time(:second) - 400`)
- Action: push `"publishFinal"` on the existing socket
- Assert: caption is processed normally and `UserTracker.get_by_key` shows a refreshed `last_publish`
- Why: post-join inactivity is NOT enforced by this issue; this test makes the decision explicit so a future issue can build on it

#### Unit tests — Absinthe subscriptions

> No existing subscription test file was found in the codebase. Create one at `test/stream_closed_captioner_phoenix_web/schema/subscriptions_test.exs`. Test Absinthe subscriptions by calling the `config` resolver function directly (unit-test the function itself) or by using `Absinthe.run/3` with the subscription document. Check `absinthe_phoenix` test helper docs for the appropriate setup pattern for this version of the library.

**Test 7: "new_twitch_caption subscription succeeds when streamer is connected (any UserTracker entry)"**
- Setup: `UserTracker.track(self(), "active_channels", channel_id, %{last_publish: 0})`; `on_exit` to untrack
- Action: invoke subscription config / run subscription document for `newTwitchCaption(channelId: channel_id)`
- Assert: config returns `{:ok, topic: channel_id}` (no error)

**Test 8: "new_twitch_caption subscription succeeds when streamer is connected but hasn't published in 300s"**
- Setup: `UserTracker.track(self(), "active_channels", channel_id, %{last_publish: System.system_time(:second) - 400})`; `on_exit` to untrack
- Action: run subscription document
- Assert: config returns `{:ok, topic: channel_id}` — **not rejected**
- Why: explicitly documents the relaxed gate semantics ("connected" not "recently active")

**Test 9: "new_twitch_caption subscription is rejected when streamer has no UserTracker entry"**
- Setup: ensure no `UserTracker` entry for `channel_id` (no setup needed, or call `untrack` to be safe)
- Action: run subscription document
- Assert: config returns `{:error, _}`
- Why: tests the nil / missing case so the code can't crash on an empty list

#### Integration test — `test/stream_closed_captioner_phoenix_web/channels/captions_channel_test.exs`

**Test 10: "caption only reaches subscriber after streamer connects their dashboard"**
- Step A: attempt `newTwitchCaption` subscription before streamer joins → assert error / rejection
- Step B: streamer `subscribe_and_join/3` for their channel
- Step C: attempt `newTwitchCaption` subscription again → assert success
- Step D: streamer pushes `"publishFinal"` with a caption
- Step E: assert the subscription receives the `newTwitchCaption` event with the correct payload

---

### Implementation order

**Cycle 1 — Add `channel_connected?/1` to `UserTracker`**
1. Write Tests 1, 2, 3 → **red** (function doesn't exist yet)
2. In `user_tracker.ex`, add:
   ```elixir
   def channel_connected?(channel_id) do
     get_by_key("active_channels", channel_id)
     |> Enum.any?()
   end
   ```
3. Tests 1, 2, 3 → **green**

**Cycle 2 — Fix `track_user/1` to set `last_publish` on join**
1. Write Tests 4 and 5 → **red** (current `track_user/1` passes `%{}`)
2. In `captions_channel.ex`, change `track_user/1`:
   ```elixir
   # Before
   UserTracker.track(self(), "active_channels", uid, %{})

   # After
   UserTracker.track(self(), "active_channels", uid, %{last_publish: System.system_time(:second)})
   ```
3. Tests 4 and 5 → **green**
4. Write Test 6 → expect **green immediately** (documents existing `handle_in` behavior; if it fails, there is a regression)

**Cycle 3 — Gate the Absinthe subscription**
1. Write Tests 7, 8, 9 → **red** (config currently accepts everything)
2. In `schema.ex`, update the `new_twitch_caption` subscription:
   ```elixir
   config fn args, _info ->
     if StreamClosedCaptionerPhoenixWeb.UserTracker.channel_connected?(args.channel_id) do
       {:ok, topic: args.channel_id}
     else
       {:error, "channel is not active"}
     end
   end
   ```
3. Tests 7, 8, 9 → **green**

**Cycle 4 — Integration and regression check**
1. Write Test 10 → expect **green** if Cycles 1–3 are correct; investigate any failures
2. Run `mix test` for the full suite:
   - `send_chat_reminder_test.exs` — calls `channel_active?` (unchanged); should pass without modification
   - Any showcase tests — Cycle 2 means channels appear "recently active" from the moment the streamer opens their dashboard (before first caption); check if expectations need updating
3. Run `mix lint` and `mix security`

---

### Files to change

| File | Change |
|------|--------|
| `lib/stream_closed_captioner_phoenix_web/channels/user_tracker.ex` | Add `channel_connected?/1` function |
| `lib/stream_closed_captioner_phoenix_web/channels/captions_channel.ex` | `track_user/1`: pass `%{last_publish: System.system_time(:second)}` instead of `%{}` |
| `lib/stream_closed_captioner_phoenix_web/schema.ex` | Update `new_twitch_caption` subscription `config` fn to gate on `UserTracker.channel_connected?/1` |
| `test/stream_closed_captioner_phoenix_web/channels/user_tracker_test.exs` | Add Tests 1, 2, 3 |
| `test/stream_closed_captioner_phoenix_web/channels/captions_channel_test.exs` | Add Tests 4, 5, 6, 10; do NOT add `async: true` |
| `test/stream_closed_captioner_phoenix_web/schema/subscriptions_test.exs` _(new)_ | Create and add Tests 7, 8, 9; do NOT add `async: true` |

---

### Edge cases and considerations

1. **`channel_connected?` vs `channel_active?` — keep both, don't merge:** `channel_active?` (300s recency) is intentionally kept as-is for `SendChatReminder` and the showcase. `channel_connected?` (any entry present) is the new predicate for the subscription gate. Test 3 documents the semantic split so future developers don't accidentally swap them.

2. **`UserTracker` cannot be mocked:** It is a supervised `GenServer` started in `Application`, not a behaviour. Tests must call the real tracker and must NOT use `async: true`. Add `on_exit` hooks to `untrack` every test-seeded entry. Follow the existing `UserTrackerTest` pattern (lines 6–8 of that file).

3. **`Phoenix.Tracker` automatic cleanup on disconnect:** When the streamer closes their dashboard, their `CaptionsChannel` process dies, and `Phoenix.Tracker` automatically removes their `UserTracker` entry. This means the gate's state is self-managing — no explicit `terminate/2` cleanup is needed in `CaptionsChannel` for the gate to work correctly for new subscriptions.

4. **Existing subscriptions when streamer disconnects (deferred):** Proactive termination is out of scope — see Revision notes. When the streamer disconnects, the `UserTracker` entry disappears and publishing stops. Existing viewer connections go idle. New subscription attempts will be rejected immediately. This is acceptable behavior for this issue.

5. **Showcase side effect from Cycle 2:** After fixing `track_user/1`, channels appear "recently active" (`channel_active? == true`) as soon as the streamer opens their dashboard, even before speaking. The showcase (which calls `recently_active_channels/0`) may surface more channels. This is likely the correct behavior ("opened dashboard = intending to caption"), but verify with showcase tests and confirm with the issue owner if the showcase behavior change is undesirable.

6. **`async: false` is the default for `ChannelCase`:** The `CaptionsChannelTest` comment notes `async: false` is required but does not explicitly set it — `ChannelCase` already defaults to it. Confirm this holds for the new subscription test file. If `async: true` is accidentally added, `UserTracker` state from one test can corrupt another.

7. **Topic key alignment verified end-to-end:** `Absinthe.Subscription.publish/3` uses `new_twitch_caption: user.uid`. The subscription `config` fn sets `topic: args.channel_id`. The issue owner confirmed `args.channel_id == user.uid`. Test 10 verifies this alignment in an end-to-end flow.

---

### Open questions

~~1. Does `args.channel_id` equal `user.uid`?~~ **Resolved: Yes — confirmed by issue owner.**

~~2. Should the Phoenix Channel `join/3` also gate on activity?~~ **Resolved: No — only the consumer Absinthe subscription is gated.**

~~3. Should the gate use `stream.offline` EventSub instead of `UserTracker`?~~ **Resolved: No. "Active" = streamer has captions dashboard open = has a `UserTracker` entry.**

4. **Follow-up issue needed — proactive subscriber termination:** When the streamer disconnects, existing viewer subscriptions stay open but idle. Terminating them requires Absinthe private internals or a per-channel subscription-ID registry. File a separate issue for this if the idle-connection behavior is unacceptable.

5. **Showcase behavior after Cycle 2:** Should the showcase show a channel as "active" from the moment the streamer opens their dashboard (before their first caption), or only after first publish? Cycle 2 changes this to "dashboard-open = active." Confirm this is the intended behavior.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by talk2MeGooseman** [comment]
> 5\. **Showcase behavior after Cycle 2:** Should the showcase show a channel as "active" from the moment the streamer opens their dashboard (before their first caption), or only after first publish? Cycle 2 changes this to "dashboard-open = active." Confirm this is the intended behavior.

First publish.
